cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,288 @@
+	// The format we should expect for ccm worker firewall names.
+	firewallWorkerCCMNameFormat = "k8s-%s-ccm"
+)
+
+// cachedFirewall stores the current state of the CCM worker firewall.
+type cachedFirewall struct {
+	// The cached state of the firewall.
+	state *godo.Firewall
+}
+
+// firewallCache stores a cached firewall and mutex to handle concurrent access.
+type firewallCache struct {
+	mu       *sync.RWMutex // protects firewallMap.
+	firewall *cachedFirewall
+}
+
+// firewallManagerOp handles communication with methods of the FirewallManager.
+type firewallManagerOp struct {
+	client             *godo.Client
+	fwCache            *firewallCache
+	workerFirewallName string
+}
+
+// FirewallManager manages the interaction with the DO Firewalls API.
+type FirewallManager interface {
+	// Get returns the current CCM worker firewall representation (i.e., the DO Firewall object).
+	Get(ctx context.Context) (*godo.Firewall, error)
+
+	// Set applies the given inbound rules to the CCM worker firewall.
+	Set(ctx context.Context, inboundRules []godo.InboundRule) error
+}
+
+// FirewallController helps to keep cloud provider service firewalls in sync.
+type FirewallController struct {
+	kubeClient      clientset.Interface
+	client          *godo.Client
+	tags            []string
+	serviceLister   corelisters.ServiceLister
+	firewallManager *FirewallManager
+	fwCache         *firewallCache
+	ccmFirewallName string
+}
+
+// NewFirewallController returns a new firewall controller to reconcile CCM worker firewall state.
+func NewFirewallController(
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	tags []string,
+	fwManager *FirewallManager,
+	ctx context.Context,
+) (*FirewallController, error) {
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		if err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage("firewall_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return nil, err
+		}
+	}
+
+	fc := &FirewallController{
+		kubeClient:      kubeClient,
+		client:          client,
+		firewallManager: fwManager,
+		fwCache: &firewallCache{
+			mu: new(sync.RWMutex),
+		},
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.onServiceChange(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.onServiceChange(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.onServiceChange(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc, nil
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		targetFirewallState := fc.fwCache.firewall.state
+		fm := *fc.firewallManager
```

Comment:
For some reason, when I do it like that. I get:
`fc.firewallManager.Get undefined (type *FirewallManager has no field or method Get)compiler` @timoreimann

Reply:
It's because the `firewallManager` field on the `FirewallController` struct is defined as a `*FirewallManager`. In Go you pretty much never take a pointer to an interface but the interface directly. If you change the type to `FirewallManager` (and similarly adjust the type in `NewFirewallController`) things should work.